### PR TITLE
Fix invalid grant bug

### PR DESCRIPTION
**Pull request by jdatskuid**
# Jira Ticket

https://inflight.atlassian.net/browse/PLIN-4534

# Description

Fixes an internal bug which caused an unrecoverable server crash if an OAuth server responded with HTML instead of JSON or formdata.

Also, fixed some basic errors which were preventing the unit tests from running (these changes were separated into their own commit for the sake of reviewing)

### CODE REVIEW

File: src/oauth-shim.js [hunk 1]
```diff
@@ -196,12 +196,34 @@ oauth_shim.redirect = function(req, res, next) {
...
-		});
+		try {
+			res.writeHead(302, {
+				'Access-Control-Allow-Origin': '*',
+				'Location': path
+			});
+
+			res.end();
+		} catch(err) {
+			res.writeHead(500, {
```
Comment (acofer) [inline src/oauth-shim.js]: should be 502

File: src/oauth-shim.js [hunk 2]
```diff
@@ -196,12 +196,34 @@ oauth_shim.redirect = function(req, res, next) {
...
+			});
 
-		res.end();
+			if (err.message === "Invalid character in header content [\"Location\"]") {
+				// If the authenticating server returned a malformed response (e.g. an HTML page) that was accidentally
+				// parsed as form data, then we might have illegal characters in the oauthshim.data object's property names
+				// or values.
+				res.end(JSON.stringify({
+					error: err.message,
+					error_message: 'The authenticating server included illegal header characters in its response (it might not have responded with an expected format)',
```
Comment (acofer) [inline src/oauth-shim.js:217]: 🙌

File: test/e2e/oauth-shim.js [hunk 3]
```diff
@@ -284,6 +284,31 @@ describe('OAuth2 exchange refresh_token for access token', function() {
...
+				<details>
+					<summary>Error occurred in authorization</summary>
+					<p>OAuth client integration with the given client id is not found.</p>
+				</details>
+			</body>
+			</html>`;
+
+		request(app)
+			.get('/proxy?' + querystring.stringify(query))
+			.expect(500)
```
Comment (acofer) [inline test/e2e/oauth-shim.js]: here too. :-)

File: src/oauth-shim.js [hunk 4]
```diff
@@ -196,12 +196,34 @@ oauth_shim.redirect = function(req, res, next) {
...
+				}));
+			} else {
+				// For any other error, don't risk leaking sensitive data to the client. Just return an unknown error.
+				res.end(JSON.stringify({
+					error: 'Unknown Error',
+					error_message: 'Unknown Error',
+				}));
+			}
+		}
```
Comment (jdatskuid) [inline src/oauth-shim.js]: This is the only change of consequence.

Note that I opted for a `500` error. I haven't really put in the work to know that the response from the OAuth server is invalid. It's just that we can't process it.

**Review by acofer — CHANGES_REQUESTED** [review]

File: src/oauth-shim.js (hunk 1 above)
Comment (jdatskuid) [reply src/oauth-shim.js]: As discussed, we are going to assume that if we've reached this point, it's definitely the authenticating server's fault for not formatting the data correctly (to spec)

**Review by acofer — APPROVED** [review]

**Pull request merged**